src/julia_gc.c

```diff
@@ -0,0 +1,755 @@
+{
+    JCache = cache;
+    JSp = sp;
+    BagHeader * hdr = (BagHeader *)obj;
+    Bag         contents = (Bag)(hdr + 1);
+    UInt        tnum = hdr->type;
+    YoungRef = 0;
+    OldObj = GcOld(obj);
+    TabMarkFuncBags[tnum]((Bag)&contents);
+    if (OldObj && YoungRef)
+        jl_gc_mark_push_remset(JuliaTLS, obj, YoungRef);
+}
+
+
+void InitBags(UInt              initial_size,
+              TNumStackFuncBags stack_func,
+              Bag *             stack_bottom,
+              UInt              stack_align)
+{
+    // HOOK: initialization happens here.
+    jl_root_scanner_hook = GapRootScanner;
+    jl_gc_disable_generational = 0;
+    jl_nonpool_alloc_hook = alloc_bigval;
+    jl_nonpool_free_hook = free_bigval;
+    for (UInt i = 0; i < NTYPES; i++)
+        TabMarkFuncBags[i] = MarkAllSubBags;
+    jl_extend_init();
+    JuliaTLS = jl_extend_get_ptls_states();
+    // jl_gc_enable(0); /// DEBUGGING
+    max_pool_obj_size = jl_extend_gc_max_pool_obj_size();
+    Module = jl_new_module(jl_symbol("ForeignGAP"));
+    datatype_mptr = jl_new_foreign_type(jl_symbol("Bag"), Module, jl_any_type,
+                                        JMarkMPtr, NULL, 1, 0);
+    datatype_bag = jl_new_foreign_type(jl_symbol("BagInner"), Module,
+                                       jl_any_type, JMarkBag, NULL, 1, 0);
+    datatype_largebag = jl_new_foreign_type(
+        jl_symbol("BagInner"), Module, jl_any_type, JMarkBag, NULL, 1, 1);
+    void * tmp = AllocateBagMemory(T_STRING, max_pool_obj_size + 1);
+    void * tmpstart = treap_find(bigvals, tmp);
+    bigval_startoffset = (char *)tmp - (char *)tmpstart;
+    GAP_ASSERT(jl_is_datatype(datatype_mptr));
+    GAP_ASSERT(jl_is_datatype(datatype_bag));
+    GAP_ASSERT(jl_is_datatype(datatype_largebag));
+    GapStackBottom = stack_bottom;
+    GapStackAlign = stack_align;
+}
+
+UInt CollectBags(UInt size, UInt full)
+{
+    // HOOK: perform a full collection
+    jl_gc_collect(1);
+    return 1;
+}
+
+void RetypeBag(Bag bag, UInt new_type)
+{
+    BagHeader * header = BAG_HEADER(bag);
+
+#ifdef COUNT_BAGS
+    /* update the statistics      */
+    {
+        UInt old_type; /* old type of the bag */
+        UInt size;
+
+        old_type = header->type;
+        size = header->size;
+        InfoBags[old_type].nrLive -= 1;
+        InfoBags[new_type].nrLive += 1;
+        InfoBags[old_type].nrAll -= 1;
+        InfoBags[new_type].nrAll += 1;
+        InfoBags[old_type].sizeLive -= size;
+        InfoBags[new_type].sizeLive += size;
+        InfoBags[old_type].sizeAll -= size;
+        InfoBags[new_type].sizeAll += size;
+    }
+#endif
+
+    header->type = new_type;
+}
+
+static inline Bag AllocateMasterPointer(void)
+{
+    // HOOK: Allocate memory for the master pointer.
+    // Master pointers require one word of memory.
+    void * result =
+        (void *)jl_gc_alloc(JuliaTLS, sizeof(void *), datatype_mptr);
+    memset(result, 0, sizeof(void *));
```

Comment:
This memset call is not necessary here. And at this point, perhaps we should just merge this function into `NewBag?